barretenberg/cpp/src/barretenberg/relations/poseidon2_external_relation.hpp

```diff
@@ -0,0 +1,121 @@
+        7, // external poseidon2 round sub-relation for fourth value
+    };
+
+    /**
+     * @brief Expression for the poseidon2 external gate.
+     * @details This relation is defined as:
+     * q_pos2 * ( (t6 - w_1_shift) + \alpha * (t5 - w_2_shift) +
+     * \alpha^2 * (t7 - w_3_shift) + \alpha^3 * (t4 - w_4_shift) ) = 0 where:
+     *      u1 := (w_1 + q_1)^5
+     *      u2 := (w_2 + q_2)^5
+     *      u3 := (w_3 + q_3)^5
+     *      u4 := (w_4 + q_4)^5
+     *      t0 := u1 + u2                                           (1, 1, 0, 0)
+     *      t1 := u3 + u4                                           (0, 0, 1, 1)
+     *      t2 := u2 + u2 + t1 = 2 * u2 + u3 + u4                   (0, 2, 1, 1)
+     *      t3 := u4 + u4 + t0 = u1 + u2 + 2 * u4                   (1, 1, 0, 2)
+     *      t4 := 4 * t1 + t3 = u1 + u2 + 4 * u3 + 6 * u4           (1, 1, 4, 6)
+     *      t5 := 4 * t0 + t2 = 4 * u1 + 6 * u2 + u3 + u4           (4, 6, 1, 1)
+     *      t6 := t3 + t5 = 5 * u1 + 7 * u2 + 1 * u3 + 3 * u4       (5, 7, 1, 3)
+     *      t7 := t2 + t4                                           (1, 3, 5, 7)
+     *
+     * @param evals transformed to `evals + C(in(X)...)*scaling_factor`
+     * @param in an std::array containing the fully extended Univariate edges.
+     * @param parameters contains beta, gamma, and public_input_delta, ....
+     * @param scaling_factor optional term to scale the evaluation before adding to evals.
+     */
+    template <typename ContainerOverSubrelations, typename AllEntities, typename Parameters>
+    void static accumulate(ContainerOverSubrelations& evals,
+                           const AllEntities& in,
+                           const Parameters&,
+                           const FF& scaling_factor)
+    {
+        using Accumulator = std::tuple_element_t<0, ContainerOverSubrelations>;
+        using View = typename Accumulator::View;
+        auto w_l = View(in.w_l);
+        auto w_r = View(in.w_r);
+        auto w_o = View(in.w_o);
+        auto w_4 = View(in.w_4);
+        auto w_l_shift = View(in.w_l_shift);
+        auto w_r_shift = View(in.w_r_shift);
+        auto w_o_shift = View(in.w_o_shift);
+        auto w_4_shift = View(in.w_4_shift);
+        auto q_l = View(in.q_l);
+        auto q_r = View(in.q_r);
+        auto q_o = View(in.q_o);
+        auto q_4 = View(in.q_4);
+        auto q_poseidon2_external = View(in.q_poseidon2_external);
+
+        // add round constants which are loaded in selectors
+        auto v1 = w_l + q_l;
+        auto v2 = w_r + q_r;
+        auto v3 = w_o + q_o;
+        auto v4 = w_4 + q_4;
+
+        // apply s-box round
+        auto u1 = v1 * v1;
+        u1 *= u1;
+        u1 *= v1;
+        auto u2 = v2 * v2;
+        u2 *= u2;
+        u2 *= v2;
+        auto u3 = v3 * v3;
+        u3 *= u3;
+        u3 *= v3;
+        auto u4 = v4 * v4;
+        u4 *= u4;
+        u4 *= v4;
+
+        // matrix mul with 14 additions
+        auto t0 = u1 + u2; // A + B
```

Comment:
Can you choose between $A, B, C, D$ or $u_1,...,u_4$ and use those consistently? I feel it hurts readability here.

Reply:
I feel it would also help clarity to rename $t_4,...,t_7$ since they are distinct and are equal to the values we are calculating. How about to $v_1,...,v_4$, so the matrix mul we are calculating is $v = M_{\mathcal{E}} u$, and also this is the order in which we put the result in `evals`?

Reply:
the A, B, C, D is consistent with the crypto implementation, but I could change it to u_i here.

Reply:
updated to u_i